src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -77,92 +78,75 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
-                process.StartInfo.EnvironmentVariables[DOTNET_CLI_UI_LANGUAGE] = "en-US";
-
-                process.OutputDataReceived += (_, e) =>
+                string filePath = Path.Combine(dir, exeName);
+                if (File.Exists(Path.Combine(dir, exeName)))
                 {
-                    if (!string.IsNullOrWhiteSpace(e.Data))
-                    {
-                        lines.Add(e.Data);
-                    }
-                };
-
-                process.Start();
-            }
-            catch
-            {
-                // when error running dotnet command, consider dotnet as not available
-                yield break;
+                    dotnetPath = filePath;
+                    break;
+                }
             }
 
-            process.BeginOutputReadLine();
-
-            process.WaitForExit();
-
-            var outputString = string.Join(Environment.NewLine, lines);
-
-            var matched = DotNetBasePathRegex.Match(outputString);
-            string basePath = null;
-            if (matched.Success)
+            if (dotnetPath != null)
             {
-                basePath = matched.Groups[1].Value.Trim();
-                yield return basePath;
+                dotnetPath = Path.GetDirectoryName(isWindows ? dotnetPath : realpath(dotnetPath) ?? dotnetPath);
             }
 
-            var lineSdkIndex = lines.FindIndex(line => line.Contains("SDKs installed"));
-
-            List<string> paths = new List<string>();
-            if (lineSdkIndex != -1)
+            string bestSDK = null;
+            int rc = NativeMethods.hostfxr_resolve_sdk2(exe_dir: dotnetPath, working_dir: workingDirectory, flags: 0, result: (key, value) =>
             {
-                lineSdkIndex++;
-
-                while (lineSdkIndex < lines.Count && !string.IsNullOrEmpty(lines[lineSdkIndex]))
+                if (key == NativeMethods.hostfxr_resolve_sdk2_result_key_t.resolved_sdk_dir)
                 {
-                    var sdkMatch = SdkRegex.Match(lines[lineSdkIndex]);
+                    bestSDK = value;
+                }
+            });
 
-                    if (!sdkMatch.Success)
-                        break;
+            if (rc == 0 && bestSDK != null)
+            {
+                yield return bestSDK;
+            }
 
-                    var version = sdkMatch.Groups[1].Value.Trim();
-                    var path = sdkMatch.Groups[2].Value.Trim();
-                    
-                    path = Path.Combine(path, version) + Path.DirectorySeparatorChar;
+            string[] paths = null;
+            NativeMethods.hostfxr_get_available_sdks(exe_dir: dotnetPath, result: (key, value) =>
+            {
+                paths = value;
+            });
 
-                    if (!path.Equals(basePath))
-                        paths.Add(path); 
-                                    
-                    lineSdkIndex++;
-                }
+            if (rc != 0)
```

Comment:
Do we need any means of error reporting here? As is we're silently ignoring failures, which might be reasonable but that isn't fully obvious to me. Justify with a comment?

Reply:
I don't fully understand this (and didn't test it, since I don't know how to introduce an error in hostfxr), but I _think_ errors are automatically printed to stderr:
https://github.com/dotnet/runtime/blob/8e407e32bbb13874a7a8b8651d6d3f1d70574754/src/native/corehost/hostfxr.h#L64

I can add a comment to that effect if you think it reasonable. This part was more aimed at doing something reasonable after writing the error. Note that the docs I looked at all checked the return value but didn't do anything with it other than returning false or whatever.

That said, this made me notice the more important bug: I wasn't resetting rc with the hostfxr_get_available_sdks call, so this part is completely invalid 🙂

Reply:
Hm. I don't love us writing to stderr accidentally. What I was wondering was whether we should throw an exception if the native call returns anything other than `0`. I think my ideal would be to overwrite that error mechanism, capture it to a string, and throw with that string. What does the SDK resolver do?

Reply:
Basically nothing. They have one replace where they assert it only fails if the directory is null:
https://github.com/dotnet/sdk/blob/a30e465a2e2ea4e2550f319a2dc088daaafe5649/src/Resolvers/Microsoft.DotNet.NativeWrapper/NETCoreSdkResolverNativeWrapper.cs#L25
and they have another case in which they don't check at all:
https://github.com/dotnet/sdk/blob/a30e465a2e2ea4e2550f319a2dc088daaafe5649/src/Resolvers/Microsoft.DotNet.NativeWrapper/NETCoreSdkResolverNativeWrapper.cs#L45

Reply:
I think, even though those are public methods, they're probably assuming some invariants from the caller.
The right thing to do would be to patch the sdk resolver.

Reply:
> they have another case in which they don't check at all:
> https://github.com/dotnet/sdk/blob/a30e465a2e2ea4e2550f319a2dc088daaafe5649/src/Resolvers/Microsoft.DotNet.NativeWrapper/NETCoreSdkResolverNativeWrapper.cs#L45

They get the errorCode, and do nothing with it - that should change.

Reply:
You're not wrong; we could do better. On the other hand, I imagine the scenarios in which the difference between "prints to stderr" and "we caught the exception and sent our own error message" really matters are few and far between, noting that the SDK resolver has been used extensively, and that code has peacefully existed for years. I also believe failure here is pretty rare; we aren't doing anything unusual or passing in arguments, so there isn't much that can go wrong.